radix-transactions/src/manifest/static_resource_movements_visitor.rs

```diff
@@ -0,0 +1,2171 @@
+use super::*;
+use crate::model::*;
+use radix_common::prelude::*;
+use radix_engine_interface::blueprints::account::*;
+use radix_engine_interface::blueprints::locker::*;
+use std::ops::*;
+use traversal::*;
+use typed_invocations::*;
+
+/// A [`ManifestInterpretationVisitor`] that statically tracks the resources in the worktop and
+/// reports the account withdraws and deposits made.
+pub struct StaticResourceMovementsVisitor {
+    // TODO: How can we do this better? How can we abstract this so that it just does the static
+    // asset movement and make the account stuff added on top?
+    account_deposits: IndexMap<ComponentAddress, Vec<AccountDeposit>>,
+    account_withdraws: IndexMap<ComponentAddress, Vec<AccountWithdraw>>,
+    /// The resource content of the worktop.
+    worktop_fungible_contents: IndexMap<FungibleResourceAddress, FungibleBounds>,
+    /// The resource content of the worktop.
+    worktop_non_fungible_contents: IndexMap<NonFungibleResourceAddress, NonFungibleBounds>,
+    /// The sources of uncertainty about the worktop.
+    worktop_uncertainty_sources: Vec<WorktopUncertaintySource>,
+    /// The buckets tracked by the by the visitor.
+    tracked_buckets: IndexMap<ManifestBucket, BucketContent>,
+}
+
+impl StaticResourceMovementsVisitor {
+    pub fn new(initial_worktop_state_is_unknown: bool) -> Self {
+        let worktop_uncertainty_sources = if initial_worktop_state_is_unknown {
+            vec![WorktopUncertaintySource::YieldFromParent]
+        } else {
+            vec![]
+        };
+        Self {
+            account_deposits: Default::default(),
+            account_withdraws: Default::default(),
+            tracked_buckets: Default::default(),
+            worktop_fungible_contents: Default::default(),
+            worktop_non_fungible_contents: Default::default(),
+            worktop_uncertainty_sources,
+        }
+    }
+
+    pub fn output(
+        self,
+    ) -> (
+        IndexMap<ComponentAddress, Vec<AccountDeposit>>,
+        IndexMap<ComponentAddress, Vec<AccountWithdraw>>,
+    ) {
+        (self.account_deposits, self.account_withdraws)
+    }
+}
+
+impl ManifestInterpretationVisitor for StaticResourceMovementsVisitor {
+    type Error<'a> = StaticResourceMovementsError<'a>;
+
+    // region:Invocation
+    fn on_start_instruction<'a>(
+        &mut self,
+        OnStartInstruction { index, effect }: OnStartInstruction<'a>,
+    ) -> ControlFlow<Self::Error<'a>> {
+        // We only care about invocations. Ignore anything that is not an invocation.
+        let ManifestInstructionEffect::Invocation { kind, args } = effect else {
```

Comment:
I wonder if there should be an `on_finish_invocation` or something, as this seems like quite a common thing to filter for. It's a zero cost abstraction hmm...

Reply:
It would be useful to have. However, I don't think we need it too much at the moment.